Refactor/types

ashwin95r opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/376)
<!-- Reviewable:end -->

manishrjain commented:
Review status: 0 of 31 files reviewed at latest revision, 43 unresolved discussions, some commit checks failed.

---

*[rdf/parse.go, line 186 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv3yV8cOYKuTcm6Jzz:-KYv3yV8cOYKuTcm6K--:b-nue9rl) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/rdf/parse.go#L186)):*
> ```Go
> 				}
> 
> 				p1 := types.ValueForType(types.BinaryID)
> ```

You don't need any of this.

---

*[types/conversion.go, line 32 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv5NmWKcIEA2L5x351:-KYv5NmWKcIEA2L5x352:b-2fbi43) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/types/conversion.go#L32)):*
> ```Go
> 	"github.com/twpayne/go-geom/encoding/wkb"
> )
> ```

type Value struct {
  id byte
  v interface{}
}

---

*[types/definitions.go, line 54 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYlAF2zpYQ82LGI14Gt:-KYlAF2zpYQ82LGI14Gu:bn8jm1v) ([raw file](https://github.com/dgraph-io/dgraph/blob/e28a55df6ef4d48b790def83575c45f57410a5d8/types/definitions.go#L54)):*
> ```Go
> // IsScalar returns true if the object is of scalar type.
> func (o Object) IsScalar() bool {
> 	if o.Id == ObjectID {
> ```

return o.Id != ObjectID

---

*[types/scalar_types.go, line 110 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYlDLz7ZTsu5JiAXdMn:-KYlDLz85luSmN_37OBb:bgdmzsj) ([raw file](https://github.com/dgraph-io/dgraph/blob/e28a55df6ef4d48b790def83575c45f57410a5d8/types/scalar_types.go#L110)):*
> ```Go
> // Type returns the type of this value
> func (v Int32) TypeID() TypeID {
> 	return Int32ID
> ```

Try to make them one line.

---

*[types/scalar_types.go, line 42 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv4td9U4Eo6i0d8knw:-KYv4td9U4Eo6i0d8knx:b-2l64mb) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/types/scalar_types.go#L42)):*
> ```Go
> 
> var typeNameMap = map[string]TypeID{
> 	"int":      Int32ID,
> ```

Can't we say stype.Posting_INT32.

---

*[types/scalar_types.go, line 122 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv48E6dphsDscmKIjh:-KYv48E6dphsDscmKIji:b-gtvain) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/types/scalar_types.go#L122)):*
> ```Go
> 
> // Type returns the type of this value
> func (v Int32) TypeID() TypeID {
> ```

Remove all these.

---

*[types/sort.go, line 62 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv4ZgV0diGHVs-3ZBP:-KYv4ZgV0diGHVs-3ZBQ:b-usbsro) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/types/sort.go#L62)):*
> ```Go
> 
> func (s byFloat) Less(i, j int) bool {
> 	return (s.values[i].(Float)) < (s.values[j].(Float))
> ```

This Float would be float64; or whatever they typedef it to.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 2 of 19 files at r2, 8 of 22 files at r3, 6 of 16 files at r4, 1 of 2 files at r5, 16 of 16 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 65 unresolved discussions, some commit checks failed.

---

*[cmd/dgraph/main.go, line 239 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvn-48pWjbd2O84ccW:-KYvn-48pWjbd2O84ccX:b-4pdv9a) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/cmd/dgraph/main.go#L239)):*
> ```Go
> 	switch v := nq.Value.Val.(type) {
> 	case *graph.Value_IntVal:
> 		i := int32(v.IntVal)
> ```

This casting should be done inside ConvertFromInterface.

---

*[geo/filter_test.go, line 37 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvnNDgyo9BqnmH0q0h:-KYvnNDgyo9BqnmH0q0i:b-4irrhm) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/geo/filter_test.go#L37)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	gd := types.ValueForType(types.StringID)
> ```

var gd string

---

*[geo/filter_test.go, line 48 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvnOjxzCb3shnQ0iNs:-KYvnOjxzCb3shnQ0iNt:b-9bo40c) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/geo/filter_test.go#L48)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	gd := types.ValueForType(types.StringID)
> ```

Do you need ValueForType?

---

*[posting/types/types.proto, line 11 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvndAoIz0ESryzCpfc:-KYvndAoIz0ESryzCpfd:bwbxbfo) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/posting/types/types.proto#L11)):*
> ```protobuf
> 	bytes value = 2;
> 	enum ValType {
> 		BINARY = 0;
> ```

Formatting.

---

*[query/query.go, line 174 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvoCtiLhVYIM5MOEkV:-KYvoCtiLhVYIM5MOEkW:b8jn87u) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/query/query.go#L174)):*
> ```Go
> 
> // getValue gets the value from the task.
> func getValue(tv *task.Value) (interface{}, types.TypeID, error) {
> ```

Create a simple Value struct, which contains interface{} and TypeID. Use that in all the new functions, and here.

---

*[query/query.go, line 182 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvp3ydSP40ZMV3chDT:-KYvp3ydSP40ZMV3chDU:b7vghru) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/query/query.go#L182)):*
> ```Go
> 		return nil, 0, x.Errorf("Invalid type: %v", vType)
> 	}
> 	x.Check(types.Convert(vID, vID, valBytes, &val))
> ```

Add a comment here.

---

*[schema/parse.go, line 220 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvpd6beyFrN06Vsm02:-KYvpd6beyFrN06Vsm03:ba3dfmt) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/schema/parse.go#L220)):*
> ```Go
> 				if t, ok := getScalar(typ); ok {
> 					if t1, ok := str[name]; ok {
> 						// TODO: Check this line
> ```

Remove.

---

*[types/conversion.go, line 34 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvpru2w4qROFBEEzNa:-KYvpru2w4qROFBEEzNb:bduickf) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L34)):*
> ```Go
> 
> // Convert converts the value to given scalar type.
> func Convert(fromID TypeID, toID TypeID, data []byte, res *interface{}) error {
> ```

Convert(data, from, to, res) -- or, move to type Value struct.

---

*[types/conversion.go, line 230 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvq3utOZirM9GtJmid:-KYvq3utOZirM9GtJmie:b7d1yw7) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L230)):*
> ```Go
> 			case Int32ID:
> 				if vc {
> 					*res = int32(1)
> ```

*res = int32(0); if vc { *res = int32(1) }

---

*[types/conversion.go, line 324 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvqCCqWqCZBjVaB50L:-KYvqCCqWqCZBjVaB50M:b-o8yer7) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L324)):*
> ```Go
> 			}
> 		}
> 	}
> ```

Have a default case.

---

*[types/conversion.go, line 328 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvqWP57e0hi_fFhqF7:-KYvqWP57e0hi_fFhqF8:bbqsvkd) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L328)):*
> ```Go
> }
> 
> func ConvertFromInterface(fromID TypeID, toID TypeID, val interface{}, res *interface{}) error {
> ```

Call it Marshal(..), and also return ([]byte, error). Then you don't need res interface{}.

Marshal(value Value, toID) -> []byte, error

---

*[types/sort.go, line 60 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvsCsYJsYXHFgWrhY6:-KYvsCsYJsYXHFgWrhY7:bmkqzl2) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/sort.go#L60)):*
> ```Go
> }
> 
> type byFloat struct{ sortBase }
> ```

We don't need all of these types, if you use the Value struct. Just one type, called byValue(..). In the Less function, do a switch case.

---

*[types/sort.go, line 73 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvrxxA2E_hkboDpZf-:-KYvrxxA2E_hkboDpZf0:ba20u2b) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/sort.go#L73)):*
> ```Go
> 
> // Sort sorts the given array in-place.
> func Sort(sID TypeID, v []interface{}, ul *task.List) error {
> ```

Sort(values []Value) -> byValue. Just one Less(i, j int) function -> switch s.values[i].typeID: s.values[i].(float64) < s.values[j].(float64)

---

*[worker/backup.go, line 41 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvrfzqMqcul9-6MO8g:-KYvrfzryccL637Y2-lQ:ba3dfmt) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/worker/backup.go#L41)):*
> ```Go
> 			str := stype.ValueForType(stype.StringID)
> 			x.Check(stype.Convert(vID, stype.StringID, p.Value, &str))
> 			fmt.Println(p.Value, vID)
> ```

Remove.

---

*[worker/sort.go, line 211 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvrijTrXW9F6kyYWhD:-KYvrijTrXW9F6kyYWhE:ba3dfmt) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/worker/sort.go#L211)):*
> ```Go
> 		}
> 		values[i] = val
> 		fmt.Println("**", val, scalar, attr)
> ```

Remove.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 14 of 32 files reviewed at latest revision, 67 unresolved discussions, some commit checks failed.

---

*[cmd/dgraph/main.go, line 239 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvn-48pWjbd2O84ccW:-KYwR-Yg-T-_UwY6F-Ex:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/cmd/dgraph/main.go#L239)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This casting should be done inside ConvertFromInterface.
</blockquote></details>

Done.

---

*[posting/types/types.proto, line 11 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvndAoIz0ESryzCpfc:-KYwRPLJLadww3nsGKP7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/posting/types/types.proto#L11)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Formatting.
</blockquote></details>

Done.

---

*[query/query.go, line 174 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvoCtiLhVYIM5MOEkV:-KYwRQY8NznVlAps_bYL:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/query/query.go#L174)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Create a simple Value struct, which contains interface{} and TypeID. Use that in all the new functions, and here.
</blockquote></details>

Done.

---

*[query/query.go, line 182 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvp3ydSP40ZMV3chDT:-KYwRcFfPOWSGqIpxDlN:bi0x5ba) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/query/query.go#L182)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add a comment here.
</blockquote></details>

Changed to val struct.

---

*[rdf/parse.go, line 186 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv3yV8cOYKuTcm6Jzz:-KYwRg41GJHt1TUKLwPI:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/rdf/parse.go#L186)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need any of this.
</blockquote></details>

Done.

---

*[schema/parse.go, line 220 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvpd6beyFrN06Vsm02:-KYvvnPdFRTNlsIXBekH:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/schema/parse.go#L220)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove.
</blockquote></details>

Done.

---

*[types/conversion.go, line 32 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv5NmWKcIEA2L5x351:-KYwRh51TMnRCIi1J9Ht:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/types/conversion.go#L32)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

type Value struct {
  id byte
  v interface{}
}

</blockquote></details>

Done.

---

*[types/conversion.go, line 34 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvpru2w4qROFBEEzNa:-KYwRiQRe_JiPZ7vaBU4:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L34)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Convert(data, from, to, res) -- or, move to type Value struct.
</blockquote></details>

Done.

---

*[types/conversion.go, line 230 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvq3utOZirM9GtJmid:-KYwSFrtBEm9oxfesyMx:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L230)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

*res = int32(0); if vc { *res = int32(1) }
</blockquote></details>

Done.

---

*[types/conversion.go, line 328 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvqWP57e0hi_fFhqF7:-KYwSoZKpsQ_fTFZV8sN:b3vl5x1) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/conversion.go#L328)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Call it Marshal(..), and also return ([]byte, error). Then you don't need res interface{}.

Marshal(value Value, toID) -> []byte, error
</blockquote></details>

using Val interface.

---

*[worker/backup.go, line 41 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvrfzqMqcul9-6MO8g:-KYvvwEiLtU4LceY3ytk:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/worker/backup.go#L41)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove.
</blockquote></details>

Done.

---

*[worker/sort.go, line 211 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvrijTrXW9F6kyYWhD:-KYvw-VOvp3mlG5F6I3n:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/worker/sort.go#L211)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 1 of 2 files at r8, 15 of 17 files at r9, 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 71 unresolved discussions, some commit checks failed.

---

*[cmd/dgraph/main.go, line 223 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwUxkEFPrgrGNiXUia:-KYwUxkEFPrgrGNiXUib:bqk40pl) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/cmd/dgraph/main.go#L223)):*
> ```Go
> 			return resp, err
> 		}
> 		if (types.Val{}) != v {
> ```

v.Value != nil

---

*[types/conversion.go, line 301 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVJdblyCuokHO2en-:-KYwVJdblyCuokHO2en0:b-vmz36r) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L301)):*
> ```Go
> 		{
> 
> 			fmt.Println("***", data)
> ```

Remove and below.

---

*[types/conversion.go, line 367 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVetVtDznxHafsSa5:-KYwVetVtDznxHafsSa6:b-gzo1um) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L367)):*
> ```Go
> 		switch toID {
> 		case StringID:
> 			*res = string(strconv.FormatInt(int64(vc), 10))
> ```

Avoid a copy here.

---

*[types/conversion.go, line 380 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVY1MQUxbKybbcrIq:-KYwVY1MQUxbKybbcrIr:bds3e0q) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L380)):*
> ```Go
> 		switch toID {
> 		case StringID:
> 			*res = string(strconv.FormatFloat(float64(vc), 'E', -1, 64))
> ```

No need to do a copy.

---

*[types/conversion.go, line 394 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVhj7fEeCMtDkaI6s:-KYwVhj7fEeCMtDkaI6t:b-80iab9) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L394)):*
> ```Go
> 		switch toID {
> 		case StringID:
> 			*res = string(strconv.FormatBool(bool(vc)))
> ```

This.

---

*[types/conversion.go, line 410 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVjFAaiVPhOdZNB43:-KYwVjFAaiVPhOdZNB44:b-80iab9) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L410)):*
> ```Go
> 		switch toID {
> 		case StringID:
> 			*res = string(vc.Format(dateFormatYMD))
> ```

This.

---

*[types/conversion.go, line 422 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVkBKRsdAyXK8ebrl:-KYwVkBKRsdAyXK8ebrm:b-80iab9) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L422)):*
> ```Go
> 		switch toID {
> 		case StringID:
> 			*res = string(vc.String())
> ```

This.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: all files reviewed at latest revision, 71 unresolved discussions, some commit checks failed.

---

*[cmd/dgraph/main.go, line 223 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwUxkEFPrgrGNiXUia:-KYwwbk08kIIWIBFOltP:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/cmd/dgraph/main.go#L223)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

v.Value != nil
</blockquote></details>

Done.

---

*[types/conversion.go, line 301 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVJdblyCuokHO2en-:-KYwxWqzlI3otPsPiV0h:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L301)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove and below.
</blockquote></details>

Done.

---

*[types/conversion.go, line 367 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVetVtDznxHafsSa5:-KYwxZ3Dq2jVDT3-YTzf:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L367)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Avoid a copy here.
</blockquote></details>

Done.

---

*[types/conversion.go, line 380 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVY1MQUxbKybbcrIq:-KYwx_4GxCQKfbJ9oHG6:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L380)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

No need to do a copy.
</blockquote></details>

Done.

---

*[types/conversion.go, line 394 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVhj7fEeCMtDkaI6s:-KYwxaXweq6fJFKRmi8C:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L394)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This.
</blockquote></details>

Done.

---

*[types/conversion.go, line 410 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVjFAaiVPhOdZNB43:-KYwxbmGFIUhc9x-Nj29:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L410)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This.
</blockquote></details>

Done.

---

*[types/conversion.go, line 422 at r10](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYwVkBKRsdAyXK8ebrl:-KYwxKfa0ohbrr2K0q5k:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/d4508dc0777463b794ff69390bcb770864b50f7f/types/conversion.go#L422)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This.
</blockquote></details>

Done.

---

*[types/sort.go, line 73 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvrxxA2E_hkboDpZf-:-KYx-1qPQorOqasCLxSA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/sort.go#L73)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Sort(values []Value) -> byValue. Just one Less(i, j int) function -> switch s.values[i].typeID: s.values[i].(float64) < s.values[j].(float64)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

jchiu0 commented:
Reviewed 1 of 19 files at r2, 1 of 22 files at r3, 3 of 16 files at r4, 3 of 16 files at r7, 8 of 17 files at r9, 2 of 3 files at r10, 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 64 unresolved discussions, some commit checks failed.

---

*[posting/index.go, line 129 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ02zJVeD3pHZ2p62ly:-KZ02zJWSJYprNTsndvm:btj56kd) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/posting/index.go#L129)):*
> ```Go
> // addIndexMutations adds mutation(s) for a single term, to maintain index.
> func addIndexMutations(ctx context.Context, attr string, uid uint64,
> 	typ byte, data []byte, del bool) {
> ```

Maybe input "typ" can be of type types.TypeID?

Maybe typID can be called dataType.

---

*[posting/index.go, line 209 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ03itycEir6XIGR_Wq:-KZ03itycEir6XIGR_Wr:bhlrlz8) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/posting/index.go#L209)):*
> ```Go
> 	if verr == nil && len(vbytes) > 0 {
> 		delTerm := vbytes
> 		/*
> ```

Delete comments

---

*[posting/index.go, line 219 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ03hNZoHFbbRYwaXv2:-KZ03hNZoHFbbRYwaXv3:bhlrlz8) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/posting/index.go#L219)):*
> ```Go
> 	if op == Set {
> 		/*
> 			p := types.ValueForType(types.TypeID(t.ValueType))
> ```

Delete comments

---

*[posting/types/types.proto, line 11 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvndAoIz0ESryzCpfc:-KZ-qwxEEOyirGNwvz69:b-80mwet) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/posting/types/types.proto#L11)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Tabs.

---

*[query/query_test.go, line 119 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ-p0GzLWXrAemaL0aZ:-KZ-p0H-lqN3L1efk4si:b-rcwaus) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/query/query_test.go#L119)):*
> ```Go
> }
> 
> /*
> ```

You are disabling a lot of tests?

---

*[rdf/parse.go, line 181 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ04GaqU05DN-e0ROYm:-KZ04GaqU05DN-e0ROYn:b-urjfwb) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/rdf/parse.go#L181)):*
> ```Go
> 				src := types.ValueForType(types.StringID)
> 				src.Value = []byte(oval)
> 				err := types.Convert(src, &p)
> ```

Small suggestion: I wonder if we can have a function like
types.Convert(types.Value, types.typeID) that will return types.Value, error.

---

*[types/date_test.go, line 21 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ016D4JxXA-vaoMwMU:-KZ016D5CYa79ssuLP-_:b-1tgfy9) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/types/date_test.go#L21)):*
> ```Go
> //	"math"
> 
> /*
> ```

Tests disabled?

---

*[types/sort.go, line 60 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvsCsYJsYXHFgWrhY6:-KZ-m5mx3kUwRaiIeN6D:b-4jgvan) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/types/sort.go#L60)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

We don't need all of these types, if you use the Value struct. Just one type, called byValue(..). In the Less function, do a switch case.
</blockquote></details>

Reason why I wrote it this way is for efficiency, especially for cheap comparisons between integers. But there is probably no need for that here, so I am ok.

---

*[worker/sort.go, line 201 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ-pZQBedmanLtwp64I:-KZ-pZQBedmanLtwp64J:biqpsmj) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/worker/sort.go#L201)):*
> ```Go
> 
> // sortByValue fetches values and sort UIDList.
> func sortByValue(attr string, ul *task.List, scalar types.TypeID) error {
> ```

Rename scalar as typ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 19 of 33 files reviewed at latest revision, 61 unresolved discussions, some commit checks failed.

---

*[posting/index.go, line 129 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ02zJVeD3pHZ2p62ly:-KZ0KhCBCL_q0GZq7UBM:bs7wumb) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/posting/index.go#L129)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Maybe input "typ" can be of type types.TypeID?

Maybe typID can be called dataType.
</blockquote></details>

Done.

---

*[posting/index.go, line 209 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ03itycEir6XIGR_Wq:-KZ0KiwDdM_NrLuaIHEc:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/posting/index.go#L209)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Delete comments
</blockquote></details>

Done.

---

*[posting/index.go, line 219 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ03hNZoHFbbRYwaXv2:-KZ0Kj_X1pv6j1WwWse4:bs7wumb) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/posting/index.go#L219)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Delete comments
</blockquote></details>

Done.

---

*[query/query_test.go, line 119 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ-p0GzLWXrAemaL0aZ:-KZ44YimDVn-2hFKL2-C:b6nd4gs) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/query/query_test.go#L119)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

You are disabling a lot of tests?
</blockquote></details>

Revived them.

---

*[types/date_test.go, line 21 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ016D4JxXA-vaoMwMU:-KZ464wPNP8LCAdgJ_Fy:b7b9fn9) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/types/date_test.go#L21)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Tests disabled?
</blockquote></details>

Will revive these soon. They seem pretty direct.

---

*[types/sort.go, line 62 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYv4ZgV0diGHVs-3ZBP:-KZ44mfHhZ7Ra_7Uk9CE:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/09bb56c417dc87241d4e60492834f25f095bd113/types/sort.go#L62)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This Float would be float64; or whatever they typedef it to.
</blockquote></details>

Done.

---

*[worker/sort.go, line 201 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ-pZQBedmanLtwp64I:-KZ0JHw7AC-k-6IL6P58:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/worker/sort.go#L201)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Rename scalar as typ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: 19 of 33 files reviewed at latest revision, 69 unresolved discussions, some commit checks failed.

---

*[posting/index.go, line 116 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4WYMqctLaOqzo2GoU:-KZ4WYMqctLaOqzo2GoV:b-hmvdnh) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/posting/index.go#L116)):*
> ```Go
> 		if s == types.DateID {
> 			return types.DateIndex(attr, &v)
> 		} else {
> ```

not else.

---

*[query/query.go, line 268 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4Xt1ttGMZQBkK7Kq5:-KZ4Xt1ttGMZQBkK7Kq6:bjcbfkf) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query.go#L268)):*
> ```Go
> 				dst.SetUID(uid)
> 			} else {
> 				globalType := schema.TypeOf(pc.Attr)
> ```

bestEffortType

---

*[query/query.go, line 269 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4XuadTBD0kFuJGWjh:-KZ4XuadTBD0kFuJGWji:b-ljs3zb) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query.go#L269)):*
> ```Go
> 			} else {
> 				globalType := schema.TypeOf(pc.Attr)
> 				schemaType := pc.Params.AttrType
> ```

Add some comments to explain the difference between global and schema type.

---

*[query/query.go, line 299 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4XaHU_NedYbneeYF9:-KZ4XaHU_NedYbneeYFA:b-r35ltf) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query.go#L299)):*
> ```Go
> 					}
> 				} else {
> 					err = types.Convert(v, &sv)
> ```

x.Check

---

*[query/query_test.go, line 165 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4YqOboDG0-xdJRzUR:-KZ4YqOcufT7BcHpe6-h:b-5r3ivl) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query_test.go#L165)):*
> ```Go
> 	addEdgeToTypedValue(t, ps, "loc", 1, types.GeoID, gData.Value.([]byte))
> 
> 	data := types.ValueForType(types.BinaryID)
> ```

This is not required. data := types.Val{ID: ..}

---

*[query/query_test.go, line 168 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4YfG8UwzQ0k01Bnve:-KZ4YfG8UwzQ0k01Bnvf:bf7wfu2) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query_test.go#L168)):*
> ```Go
> 	intD := types.ValueForType(types.Int32ID)
> 	src.Value = []byte("15")
> 	err = types.Convert(src, &intD)
> ```

types.Convert(types.Val{ID: Int32ID, value: ...}, &intD)

---

*[rdf/parse.go, line 181 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ04GaqU05DN-e0ROYm:-KZ05_CGgG_yIaMBgHS0:bl80nja) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/rdf/parse.go#L181)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Small suggestion: I wonder if we can have a function like
types.Convert(types.Value, types.typeID) that will return types.Value, error.
</blockquote></details>

I think that wouldn't be any better than the current `Convert(from Value, to *Value) error`; though of course, this is subjective. I find this signature a bit better.

---

*[types/date_test.go, line 30 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4ZX2Cm41oc2bq9W1H:-KZ4ZX2Cm41oc2bq9W1I:b-zaf54j) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/types/date_test.go#L30)):*
> ```Go
> }
> 
> /*
> ```

Add a task to fix these tests after pushing the PR.

---

*[types/geo_test.go, line 31 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4Zh10E22k1xcm0xOj:-KZ4Zh10E22k1xcm0xOk:boejlb7) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/types/geo_test.go#L31)):*
> ```Go
> 	for _, v := range array {
> 		g := ValueForType(GeoID)
> 		src := ValueForType(StringID)
> ```

src := types.Val{ID: .., Value: []byte(v)}

---

*[types/sort.go, line 45 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4_-sQaVwJVy2vOIzV:-KZ4_-sQaVwJVy2vOIzW:b50di1) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/types/sort.go#L45)):*
> ```Go
> // Less compares two elements
> func (s byValue) Less(i, j int) bool {
> 	switch s.values[i].Value.(type) {
> ```

switch s.values[i].ID {

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

manishrjain commented on posting/index.go:
```diff
@@ -88,38 +90,46 @@ func initIndex() {
-		return types.TimeIndex(attr, v)
-	case *types.String:
-		return types.DefaultIndexKeys(attr, v), nil
+	switch v := sv.Value.(type) {
```
This switch case should also use typeID.

ashwin95r replied:
Have a task to clean this.

manishrjain commented on query/proto.go:
```diff
@@ -19,31 +19,33 @@ package query
-	case *types.Bytes:
-		return &graph.Value{&graph.Value_BytesVal{*val}}
+func toProtoValue(v types.Val) *graph.Value {
+	switch val := v.Value.(type) {
```
And here. Switch to TypeID.

ashwin95r replied:
Done.

manishrjain commented on query/query.go:
```diff
@@ -260,26 +255,29 @@ func (sg *SubGraph) preTraverse(uid uint64, dst outputNode) error {
-				txt, err := v.MarshalText()
+				txt := types.ValueForType(types.StringID)
+				err := types.Convert(v, &txt)
+				//txt, err := v.MarshalText()
```
remove comment.

manishrjain commented on query/query.go:
```diff
@@ -307,8 +311,8 @@ func (sg *SubGraph) preTraverse(uid uint64, dst outputNode) error {
 
-func createProperty(prop string, v types.Value) *graph.Property {
-	pval := toProtoValue(v)
+func createProperty(prop string, v interface{}) *graph.Property {
```
Why not pass v types.Val?

ashwin95r replied:
Done.

manishrjain commented on query/query.go:
```diff
@@ -754,7 +758,7 @@ func (sg *SubGraph) applyOrderAndPagination(ctx context.Context) error {
 // outputNode is the generic output / writer for preTraverse.
 type outputNode interface {
-	AddValue(attr string, v types.Value)
+	AddValue(attr string, v interface{})
```
AddValue(attr string, v types.Val)

ashwin95r replied:
Done.

manishrjain commented on query/query.go:
```diff
@@ -768,7 +772,7 @@ type protoOutputNode struct {
 
 // AddValue adds an attribute value for protoOutputNode.
-func (p *protoOutputNode) AddValue(attr string, v types.Value) {
+func (p *protoOutputNode) AddValue(attr string, v interface{}) {
```
types.Val

manishrjain commented on types/index.go:
```diff
@@ -36,7 +37,7 @@ func DefaultIndexKeys(attr string, val *String) []string {
 
 // IntIndex indexs int type.
-func IntIndex(attr string, val *Int32) ([]string, error) {
+func IntIndex(attr string, val *int32) ([]string, error) {
```
Just do Index and pass types.Val, so based on the typeid, it would do the right thing.

ashwin95r replied:
Have a task clean it.

ashwin95r commented:
Review status: 18 of 33 files reviewed at latest revision, 76 unresolved discussions, some commit checks failed.

---

*[posting/index.go, line 116 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4WYMqctLaOqzo2GoU:-KZ4bqfaEgBFAcP06okz:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/posting/index.go#L116)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

not else.
</blockquote></details>

Done.

---

*[posting/types/types.proto, line 11 at r8](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KYvndAoIz0ESryzCpfc:-KZ4cD9pn2xxNICJKxR-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c912b3045157911fb3a1c9c08eaab8d89ded806e/posting/types/types.proto#L11)):*
<details><summary><i>Previously, jchiu0 (Jay Chiu) wrote…</i></summary><blockquote>

Tabs.
</blockquote></details>

Done.

---

*[query/query.go, line 268 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4Xt1ttGMZQBkK7Kq5:-KZ4cxi5HRyIRtADT4nZ:b-277h9n) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query.go#L268)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

bestEffortType
</blockquote></details>

Added comments.

---

*[query/query.go, line 269 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4XuadTBD0kFuJGWjh:-KZ4czT6lkLF0o-UW-AT:b-277h9n) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query.go#L269)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add some comments to explain the difference between global and schema type.
</blockquote></details>

Added comments.

---

*[query/query.go, line 299 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4XaHU_NedYbneeYF9:-KZ4dFHWV7LvDRUaFjeb:bs7wumb) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query.go#L299)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

x.Check
</blockquote></details>

Done.

---

*[query/query_test.go, line 165 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4YqOboDG0-xdJRzUR:-KZ4aFdXhiR_tV-lUyNg:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query_test.go#L165)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This is not required. data := types.Val{ID: ..}
</blockquote></details>

Done.

---

*[query/query_test.go, line 168 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4YfG8UwzQ0k01Bnve:-KZ4aGiuEW34wKFYJZcx:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/query/query_test.go#L168)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

types.Convert(types.Val{ID: Int32ID, value: ...}, &intD)
</blockquote></details>

Done.

---

*[types/date_test.go, line 30 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4ZX2Cm41oc2bq9W1H:-KZ4ag3RROEF8z7JbIUu:bs7wumb) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/types/date_test.go#L30)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add a task to fix these tests after pushing the PR.
</blockquote></details>

Done.

---

*[types/geo_test.go, line 31 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4Zh10E22k1xcm0xOj:-KZ4b-J-Ugm7hFJsn0d-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/types/geo_test.go#L31)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

src := types.Val{ID: .., Value: []byte(v)}
</blockquote></details>

Done.

---

*[types/sort.go, line 45 at r12](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ4_-sQaVwJVy2vOIzV:-KZ4bWaQ4DwNEw1AapH4:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00c71b027bd2a397525b60b0545a2896b7aa8eca/types/sort.go#L45)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

switch s.values[i].ID {
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376)*
<!-- Sent from Reviewable.io -->

jchiu0 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the huge refactoring.

---

Review status: 18 of 33 files reviewed at latest revision, 72 unresolved discussions, some commit checks failed.

---

*[rdf/parse.go, line 181 at r11](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-KZ04GaqU05DN-e0ROYm:-KZ5VtNA6hT0Le23xz3V:b-pbdqz3) ([raw file](https://github.com/dgraph-io/dgraph/blob/3d7f6830e5271c9cc6fec466a10c1cce4b701007/rdf/parse.go#L181)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

I think that wouldn't be any better than the current `Convert(from Value, to *Value) error`; though of course, this is subjective. I find this signature a bit better.
</blockquote></details>

Agreed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-:-KZ5Wh5HABbc6t2_okdz:bqd642d)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Thanks for the review!

---

Review status: 18 of 33 files reviewed at latest revision, 72 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/376#-:-KZKHYK09qN47iJD34C2:b-9qdoz3)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Merged.
